core/src/validator.rs

```diff
@@ -1033,24 +1026,43 @@ impl Validator {
                 prioritization_fee_cache.clone(),
             )?;
 
-            (
-                Some(json_rpc_service),
+            let pubsub_service = if !config.rpc_config.full_api {
+                None
+            } else {
+                let (trigger, pubsub_service) = PubSubService::new(
+                    config.pubsub_config.clone(),
+                    &rpc_subscriptions,
+                    rpc_pubsub_addr,
+                );
+                config
+                    .validator_exit
+                    .write()
+                    .unwrap()
+                    .register_exit(Box::new(move || trigger.cancel()));
+
+                Some(pubsub_service)
+            };
+
+            let (completed_data_sets_sender, completed_data_sets_service) =
                 if !config.rpc_config.full_api {
```

Comment:
Technically, we could be even more restrictive and only enable this when we know RPC subscriptions are enabled. However, this isn't apparently obvious and bubbles down to the value of `--rpc-pubsub-notification-threads`. If 0, then several things turn off which effectively disable pub sub:
https://github.com/anza-xyz/agave/blob/7c12f038dd7829ce3216bdd91b5de9d24631e279/rpc/src/rpc_subscriptions.rs#L635
https://github.com/anza-xyz/agave/blob/7c12f038dd7829ce3216bdd91b5de9d24631e279/rpc/src/rpc_subscriptions.rs#L673

In a subsequent refactor, I'd like to make `RpcSubscriptions::new()` return an `Option<Self>` and then have all the pub-sub related services created only if that is a `Some(_)`. For now, the change as-is will still be a win as every node NOT using `--full-rpc-api` will avoid the overhead of this service